src/ray/gcs/gcs_server/gcs_health_check_manager.h

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <grpcpp/grpcpp.h>
+
+#include "absl/container/flat_hash_map.h"
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/id.h"
+#include "ray/common/ray_config.h"
+#include "src/proto/grpc/health/v1/health.grpc.pb.h"
+
+class GcsHealthCheckManagerTest;
+
+namespace ray {
+namespace gcs {
+
+/// GcsHealthCheckManager is used to track the healthiness of the nodes in the ray
+/// cluster. The health check is done in pull based way, which means this module will send
+/// health check to the raylets to see whether the raylet is healthy or not. If the raylet
+/// is not healthy for certain times, the module will think the raylet dead.
+class GcsHealthCheckManager {
```

Comment:
protocol wise, we could to document the behavior after node failed.
i.e. when failure detected, the node_death_callback will be fired once and the failed node will be stopped checked?  Can you add the failed node again after it failed?